src/Avalonia.Base/PropertyStore/EffectiveValue`1.cs

```diff
@@ -0,0 +1,260 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using Avalonia.Data;
+
+namespace Avalonia.PropertyStore
+{
+    /// <summary>
+    /// Represents the active value for a property in a <see cref="ValueStore"/>.
+    /// </summary>
+    /// <remarks>
+    /// Stores the active value in an <see cref="AvaloniaObject"/>'s <see cref="ValueStore"/>
+    /// for a single property, when the value is not inherited or unset/default.
+    /// </remarks>
+    internal sealed class EffectiveValue<T> : EffectiveValue
+    {
+        private T? _baseValue;
+        private UncommonFields? _uncommon;
+
+        public EffectiveValue(
+            AvaloniaObject owner,
+            StyledPropertyBase<T> property,
+            T value, 
+            BindingPriority priority)
+        {
+            Value = value;
+            Priority = priority;
+            BasePriority = BindingPriority.Unset;
+
+            if (property.HasCoercion &&
+                property.GetMetadata(owner.GetType()) is { } metadata &&
+                metadata.CoerceValue is { } coerce)
+            {
+                _uncommon = new()
+                {
+                    _coerce = coerce,
+                    _uncoercedValue = value,
+                    _uncoercedBaseValue = value,
+                };
+
+                value = coerce(owner, value);
+            }
+        }
+
+        /// <summary>
+        /// Gets the current effective value.
+        /// </summary>
+        public new T Value { get; private set; }
+
+        public override void SetAndRaise(
+            ValueStore owner,
+            IValueEntry value, 
+            BindingPriority priority)
+        {
+            Debug.Assert(priority != BindingPriority.LocalValue);
+            UpdateValueEntry(value, priority);
+            SetAndRaiseCore(owner, (StyledPropertyBase<T>)value.Property, (T)value.GetValue()!, priority);
```

Comment:
This `(T)value.GetValue()!` is  boxing all the time. Shows up as a major GC point when profiling in larger apps.

Reply:
Committed a fix for this.